include/fsm/fsm.h

```diff
@@ -156,8 +156,8 @@ fsm_getopaque(struct fsm *fsm, const struct fsm_state *state);
  * Returns NULL on error, or if there is no state.
  */
 struct fsm_state *
-fsm_collate(struct fsm *fsm,
-	int (*predicate)(const struct fsm *, const struct fsm_state *));
+fsm_collate(struct fsm *fsm, void *pred_arg,
+	int (*predicate)(void *, const struct fsm *, const struct fsm_state *));
```

Comment:
Name it `void *opaque` please, and (sorry to be annoying), move it to the last argument for `(*predicate)()`